### ISSUE TITLE
For NCEDense, BCE decreases, SoftmaxCE increases

### ISSUE THREAD

**Issue by zjost** [dmlc/gluon-nlp#684]
I have implemented something like the following example in the docs:
```
# network with sampling for training
encoder = Encoder(..)
decoder = NCEDense(..)
train_net.add(encoder)
train_net.add(decoder)
loss_train = SigmoidBinaryCrossEntropyLoss()

# training
for x, y, sampled_values in train_batches:
    pred, new_targets = train_net(x, sampled_values, y)
    l = loss_train(pred, new_targets)

# network for testing
test_net.add(encoder)
test_net.add(Dense(..., params=decoder.params))
loss_test = SoftmaxCrossEntropyLoss()

# testing
for x, y in test_batches:
    pred = test_net(x)
    l = loss_test(pred, y)
```
Where I train by providing context inputs (x) and either true or fake targets (y).  The BCE decreases for both the training and holdout set.

However, if I add a Dense layer using the decoder params and calculate the SoftmaxCE, this value increases with time for both the training and holdout sets.

How can it be that the network learns to discriminate between real and noise (context, target) pairs, but then using the NCEDense params in a dense layer to create input into Softmax gets worse?

**Comment by eric-haibin-lin** [comment]
@zjost thanks for posting the code sample. Are you saying that in this example `loss_train ` keeps decreasing but `loss_test ` increases over time? What sampler are you using to generate negatives?

**Comment by zjost** [comment]
@eric-haibin-lin that's exactly right:  `loss_train` decreases for both training and holdout datasets (so it doesn't seem to be an overfitting problem), but `loss_test` increases for both training and holdout datasets.  

I had to implement my own sampler because I couldn't find any pre-made implementations that generated the tuple of three tensors needed for `sampled_values`.  So I made a thin wrapper around the [UnigramCandidateSampler](https://github.com/dmlc/gluon-nlp/blob/master/src/gluonnlp/data/candidate_sampler.py#L30) that returns those sample candidates, along with the empirical probabilities for each of the true and sampled classes, as observed in the training data.

**Comment by eric-haibin-lin** [comment]
@zjost Got it. Are you returning the probability of candidates, or the expected count of candidates? The NCEBlock expects the expected count as inputs. Can you try returning `probability*num_candidate` instead?

Another sampler you can try is the `LogUniformSampler ` in https://github.com/dmlc/gluon-nlp/blob/master/scripts/language_model/sampler.py#L24 . You can find an example usage here: https://github.com/dmlc/gluon-nlp/blob/master/scripts/language_model/large_word_language_model.py 
Note that the LogUniformSampler expects word ids in sorted order based on their frequencies.

**Comment by zjost** [comment]
@eric-haibin-lin I have re-implemented everything on a new, simpler problem and am getting the same behavior.  I have made the correction to return `sample_probability*num_samples` for the `expected_count_sampled` and `expected_count_true` in my sampler, and have verified that the samples I'm getting approximate the actual global distribution.

As a comparison, I also trained a network with a Dense layer decoder and directly minimized the Softmax loss function without using any kind of negative sampling.  On the same dataset, this loss decreases as expected.  This at least shows that the learning task is possible on this data.

**Comment by zjost** [comment]
I have worked on this more and discovered that if I allow training to keep running, eventually the SoftMaxCE *will* start to improve.  It seems that the loss curve always increases for the first several epochs, but eventually starts reducing.

I'm still curious to understand why this happens, and if there are better ways to e.g. schedule the learning rate to get improved convergence.  However, I don't think there's an issue with the code/implementation, so this issue can be closed.  I apologize for the false alarm.
